### ISSUE TITLE
docs: clarify the meaning of “Time”

### ISSUE THREAD

**Issue by bionicles** [google/jax#8599]
Please:

- [x] Check for duplicate requests.
- [x] Describe your goal, and if possible provide a code snippet with a motivating example.

This grew oddly specific, but really I just want to make a generalized function benchmark, we give it data and functions and a maximum timeout, it gives us the grid of results = function(data) and elapsed times, and stops the functions if they run longer than the maximum timeout. Could be cool if we could jit-compile that sort of "outer loop" since longer function runtimes would seem worse for users and the environment and faster benchmarks could also yield better feedback on performance tuning both Jax and stuff written with Jax. 

Inspired by https://github.com/google/jax/discussions/8477 with @jakevdp - Here's how it might look. 1st benchmark is python. 2nd benchmark w/ jit doesn't work now but i wish i could figure out how to make it work)

How could we make it easier to compile benchmarks like these and work with elapsed time / function timeouts inside jax.jit?

```py
"goal: play with compiled benchmarks for compiled functions in jax"
from functools import partial
from itertools import product
from typing import Callable, List, Tuple
from time import perf_counter

from func_timeout import func_timeout
from pydantic import BaseSettings
import pandas as pd
import jax

jnp = jax.numpy
jit = jax.jit

# config
class Environment(BaseSettings):
    n_x: List[int] = [4, 8, 16, 32, 64, 128, 256, 512, 1024]
    min_weight: float = -1.0
    max_weight: float = 1.0
    n_trials: int = 10
    timeout_seconds: int = 30
    seed: int = 42


env = Environment()

# types
PRNGKey = jax._src.random.PRNGKey
Array = jnp.ndarray


# functions - 'benchmark'
def benchmark_no_jit(
    x: Array,
    functions: List[Callable],
    timeout_seconds: int = env.timeout_seconds,
    min_weight: float = env.min_weight,
    max_weight: float = env.max_weight,
) -> Tuple[Array, Array]:
    n_x = x.shape[0]
    n_functions = len(functions)
    outputs = jnp.empty((n_functions + 1, n_x)).at[-1, :].set(1.0)
    timings = jnp.empty(n_functions)
    for i, fun in enumerate(functions):
        try:
            # Without JIT, we can use func_timeout to set timeouts
            # and measure runtime with time.perf_counter
            start = perf_counter()
            y = func_timeout(timeout_seconds, fun, args=(x,))
            elapsed = perf_counter() - start
            y = y if isinstance(y, Array) and y.shape[0] == n_x else 0
        except Exception as e:
            print(e)
            y, elapsed = 0, timeout
        outputs = outputs.at[i, :].set(y)
        timings = timings.at[i].set(elapsed)
    outputs = jnp.nan_to_num(
        jnp.clip(outputs, a_min=min_weight, a_max=max_weight)
    ).transpose()
    return outputs, timings


# something like this...
@partial(jit, static_argnums=(0,1))
def _benchmark_with_jit(
    n_x: int,
    n_functions: int,
    x: Array,
    functions: List[Callable],
    timeout_seconds: int = env.timeout_seconds,
    min_weight: float = env.min_weight,
    max_weight: float = env.max_weight,
) -> Tuple[Array, Array]:
    outputs = jnp.empty((n_functions + 1, n_x)).at[-1, :].set(1.0)
    timings = jnp.empty(n_functions)
    for i, fun in enumerate(functions):
        try:
            with jax.profiler.trace() as trace:
                # 1: set function timeouts / time limits (otherwise, search time is infinite)
                y = trace.timeout(timeout_seconds, fun, args=(x,))
                # 2: measure how long functions take to run (to find faster functions)
                elapsed = trace.elapsed()
                y = y if isinstance(y, Array) and y.shape[0] == n_x else 0
        except Exception as _:
            y, elapsed = 0, timeout
        outputs = outputs.at[i, :].set(y)
        timings = timings.at[i].set(elapsed)
    outputs = jnp.nan_to_num(
        jnp.clip(outputs, a_min=min_weight, a_max=max_weight)
    ).transpose()
    return outputs, timings


def f1(x: Array) -> Array:
    return x ** 2 - 1


def f2(x: Array) -> Array:
    return x / 2


def f3(x: Array) -> Array:
    return x ** 3 * jnp.sin(x)


functions = [jit(f) for f in (f1, f2, f3)]

# because we can
def benchmark_benchmark(
    benchmark: Callable = benchmark_no_jit,
    words: str = "Python + func_timeout + time.perf_counter",
    functions: List[Callable] = functions,
    N_x: List[int] = env.n_x,
    seed: int = env.seed,
) -> float:
    n_functions = len(functions)
    begin = perf_counter()
    rng = PRNGKey(seed)
    df = []  # i admit to sucking at pandas
    for i, n_x in product(range(env.n_trials), N_x):
        rng, key = jax.random.split(rng)
        x = jax.random.normal(key, (n_x,))
        if isinstance(benchmark, dict):
            print("n_x", n_x)
            benchmark = benchmark[n_x]
            print("benchmark", benchmark)
            print("x", x)
        outputs, timings = benchmark(x, functions)
        # print(timings)
        # print(outputs.shape)
        # print(i, n_x, timings)
        assert outputs.shape == (n_x, n_functions + 1)
        assert timings.shape == (n_functions,)
        assert (timings > 0).all()
        assert (timings < env.timeout).all()
        assert (env.min_weight <= outputs).all()
        assert (env.max_weight >= outputs).all()
        # print(f"All assertions passed.")
        df.append(
            {
                "n_x": n_x,
                "i": i,
                **{f"f{i+1} elapsed": t for i, t in enumerate(timings)},
            }
        )
    df = pd.DataFrame(df).set_index(["n_x", "i"])
    elapsed = perf_counter() - begin
    print(df)
    print(f"{words} took {elapsed:.2f} seconds")
    return elapsed


def test_jax_benchmarks():
    elapsed_no_jit = benchmark_benchmark()
    elapsed_with_jit = benchmark_benchmark(
        benchmark={
            n: partial(_benchmark_with_jit, n_x=n, n_functions=len(functions))
            for n in env.n_x
        },
        words="XLA (jax.jit) + jax.profiler.trace",
    )
    print("done!")


if __name__ == "__main__":
    test_jax_benchmarks()
```

**Comment by hawkinsp** [comment]
Assigning to Jake to take a look, but my gut feeling is this kind of API doesn't have a sufficiently good power-to-weight ratio to be a public JAX API.

**Comment by bionicles** [comment]
I wanted to fiddle with the algo called [Levin Search](https://www.google.com/search?q=levin+search), which is allegedly a good way to do this other thing, [Solomonoff Induction](https://www.google.com/search?q=solomonoff+induction) 

As I understand it, Solomonoff Induction is about finding probability distributions, and since you can represent distributions with functions, you want to find simple functions to explain your data. The obvious way is Levin Search, which is just starting from the simplest functions and working your way to the most complicated. First function that solves your problem wins.  

However, if some functions in your search space can run forever (and they will, if your model can implement recursion / looping / etc), then brute-force search over function space takes infinite time. That's why you *need* a timer if you want to implement Levin Search (and thus, Solomonoff Induction) Likewise, beyond my fiddlings, speed is a virtue, both speed of optimization itself and speed of that which is optimized.

The difference between "outer loop python" and "outer loop jax" is huge...you guys did a great job, jax works great, I just reached a point where, all i need to make my project run much faster and on TPUs and stuff, is an explicit timer type I can pass to the pure jitted functions to signal the start of a clock variable inside the closure of the pure compiled function, and for the clock variable to have checkpoints to time stuff (`before = checkpoint(timer); y=f(x); elapsed = checkpoint(timer)-before;`), and a way cancel functions running past a time limit.  Then the slowest part of my code (running batches of timed benchmarks in a loop) could compile to XLA. I wanted to access the jax.profiler.trace data as a noob hack to get time data. My bad if that's a dumb / low yield idea. 

@jakevdp I just wanted to give a good usage example, but it did seem pretty handy, please feel free to use it or not. What if we put experiments / notebooks / play like that in a creative/fun community jax_addons repo?

No rush or worry to get it done or do it ever. Just want to write down my reasoning clearly, and even though I spent some time on this response, I'm almost surely wrong about a lot of it. Also sorry I took 2 days to reply, it's because I already fell down a different rabbit hole in the interim. Therefore, there's zero pressure and we can revisit this whatever times somebody feels interested

TLDR: this is a feature request for timers in jit
```
from jax import numpy as jnp
class JaxProfilerTimer:
    pass
def timed_f(xs: jnp.ndarray, timer: JaxProfilerTimer) -> Tuple[jnp.ndarray, JaxProfilerTimer]:
```

Thank you

**Comment by bionicles** [comment]
You were right @hawkinsp @jakevdp and I was wrong. Please do not worry about measuring time. What is time if not a pattern, in the Now, of the concept of change adding up? But why would it exist if it isn’t real? Time doesn’t exist except in an imaginary 3D space (my consciousness topos), but imaginary and real are simply my concepts of binary falsehood and trueness, then why can’t time be both imaginary, and real? And if I care about time, doesn’t that make it worth measuring with mathematics? Maybe, maybe not. How would you redefine times in terms of lengths of geodesics through the meta-topos?

![95E9901E-F608-4E98-91E9-8F333EFA3BD6](https://user-images.githubusercontent.com/24532336/144885314-122a7ea6-b8eb-4dcd-93c3-66950db6c893.jpeg)

What do you guys think about this neo-Socratic, neo-Hegelian, 100% work-relevant, descent into madness regarding the unity of human knowledge? How do we measure time, if time doesn’t exist? If time doesn’t exist, why the heck do we measure the speed of light in seconds? If time doesn’t exist, what the heck is a past and a future?

philosophy = thesis, antithesis -> synthesis

(A Perfect Atomic Logic Universe Theory)[https://drive.google.com/drive/folders/1OyLMDE4YbFTFPDjnjkSWl8ZykeVR3obi]

how could we all think time is important if it doesn’t exist? Perhaps aren’t thinking clearly about how time and space are separate categories with different relativistic properties due to their position in the hierarchy of emergent “math voodoo”

maybe the universe needs to prove itself, in the simplest way, on the simplest problem, at infinite recursive scale, because something and nothing are differences, and difference cannot exist except relative to others, and one cannot become a stable thing, unless one contains triangles of communication, which can loopily unroll ternary logic as fast as possible, and thus, everything, both space and atoms, is simultaneously more simple AND complex than we appreciated?

were newton and Einstein grasping for new math words for something like “recursive ternary calculus of elective affinities” ?

Think about it. We’re grasping at The Mechanism here. There is no human word for that. It must be both a singular thing, and an infinity, or else it just wouldn’t make sense. “Fractal Relativistic Complex Ternary Information Calculus” are big words to us, but the universe has the simplest possible words, because everything has to be made from some bedrock particle, right? That’s not complicated. That’s simple. We could make big words all day and it wouldn’t be the same thing as The Mechanism. Even if the inner workings are complicated, it’s all complicated relative to simpler stuff inside itself. But what’s inside it? Itself!

TLDR: Is “Stuff” built from a particle, an atom, a space, a logic, physics, maths, bayes theorem, geometry, a philosophy, or all of those things, and more, happening all sorts of ways inside itself all at once?

**Comment by jakevdp** [comment]
This seems to be getting off-topic.

**Comment by bionicles** [comment]
Yeah no problemo, it’s a lot. Just stuff to think about going forward: what if we reverse engineer the logic of the universe and implement it with Jax?

**Comment by bionicles** [comment]
Lots of relevant stuff out there, but this looked like a formal treatment: https://reddit.com/r/CategoryTheory/comments/ra6mm8/sharifipour_yousefi_mathematical_morphology_via/

**Comment by bionicles** [comment]
All I mean to say is, it’s shocking to realize how truly ignorant humanity really is. We need to rethink all of the things, right?
